BitwardenShared/Core/Platform/Services/EventService.swift

```diff
@@ -0,0 +1,114 @@
+import Foundation
+import OSLog
+
+// MARK: - EventService
+
+/// A protocol for a service that manages organization events. This includes saving events to disk
+/// and uploading those events.
+///
+protocol EventService {
+    /// Save an event to disk for future upload.
+    ///
+    /// - Parameters:
+    ///   - eventType: The event to track.
+    ///   - cipherId: The ID of the relevant cipher for some events.
+    ///   - uploadImmediately: If `true` then immediately attempts an upload after saving.
+    func collect(eventType: EventType, cipherId: String?, uploadImmediately: Bool) async throws
+}
+
+extension EventService {
+    func collect(
+        eventType: EventType,
+        cipherId: String? = nil,
+        uploadImmediately: Bool = false
+    ) async throws {
+        try await collect(eventType: eventType, cipherId: cipherId, uploadImmediately: uploadImmediately)
+    }
+
+    /// Save an event to disk for future upload. Takes an error reporter to do the default handling of errors,
+    /// which is to log them with the error reporter.
+    func collect(
+        eventType: EventType,
+        cipherId: String? = nil,
+        uploadImmediately: Bool = false,
+        errorReporter: ErrorReporter
+    ) async {
+        do {
+            try await collect(eventType: eventType, cipherId: cipherId, uploadImmediately: uploadImmediately)
+        } catch {
+            errorReporter.log(error: error)
```

Comment:
> I was uncertain of where best to draw boundaries with wrapper functions, most notably because logging the events is async throws; I settled on a wrapper that does a default handling of the error (logging it) but not a wrapper that puts things in a Task. It seemed reasonable to me that we'd want to maintain the understanding that logging an event is async at the call site, but save the 4 lines of boilerplate to catch an error and log it. I'm open to discussion on this, however.

👍 I agree with your comments, and like the idea of just logging the error here instead of at every callsite.

❓ Is there a benefit of passing around the error reporter vs injecting it into `EventService`?

Reply:
The main reason for injection is so that it works better with the protocol (rather than having one throwing and one non-throwing function that do the same thing). But if we're comfortable just always catching the error and logging it, I can simplify that a lot so as not to inject it on the call. I don't think we'd ever want to do anything different if there is an error collecting an event, now that I think about it. Definitely something that can be simplified.